File: rust/main/features/config/tests/config_int_test.rs
//! Integration tests for disk-cleaner-config crate.
//! Tests TomlConfig loading, profile creation, and project scanning.

use std::fs;
use std::io::Write;
use std::path::Path;

use disk_cleaner_config::api::{ConfigProvider, ProjectScanner};
use disk_cleaner_config::{create_scanner, dir_size_bytes, format_size, load_config, load_profile};

// ─── Helpers ─────────────────────────────────────────────────────────────────

fn write_file(root: &Path, relative: &str, content: &str) {
    let full = root.join(relative);
    if let Some(parent) = full.parent() {
        fs::create_dir_all(parent).unwrap();
    }
    fs::write(&full, content).unwrap();
}

fn create_test_config(dir: &Path) -> std::path::PathBuf {
    let config_path = dir.join("profiles.toml");
    fs::write(
        &config_path,
        r#"
[settings]
default_path = ""
default_profiles = ["rust"]

[profiles.rust]
name = "Rust (Cargo)"
marker = "Cargo.lock"
type = "command"
command = "cargo clean"
clean_dir = "target"
output_pattern = 'Removed (\d+) files'
source_extensions = [".rs", ".toml"]
search_exclude = ["target"]

[profiles.node]
name = "Node.js"
marker = "package-lock.json"
alt_markers = ["yarn.lock", "pnpm-lock.yaml"]
type = "remove"
targets = ["node_modules"]
optional_targets = [".next", "dist"]
source_extensions = [".js", ".ts"]
search_exclude = ["node_modules", ".next"]
"#,
    )
    .unwrap();
    config_path
}

// ─── TomlConfig Integration ─────────────────────────────────────────────────

/// @covers: DefaultConfigProvider::load
#[test]
fn test_load_config_reads_all_profiles() {
    let tmp = tempfile::tempdir().unwrap();
    let config_path = create_test_config(tmp.path());

    let config = load_config(&config_path).unwrap();
    assert_eq!(config.profile_keys().len(), 2);
    assert!(config.profile_keys().contains(&"rust".to_string()));
    assert!(config.profile_keys().contains(&"node".to_string()));
}

/// @covers: DefaultConfigProvider::get_value
#[test]
fn test_get_value_returns_unquoted_string() {
    let tmp = tempfile::tempdir().unwrap();
    let config_path = create_test_config(tmp.path());

    let config = load_config(&config_path).unwrap();
    assert_eq!(config.get_value("profiles.rust.name"), "Rust (Cargo)");
    assert_eq!(config.get_value("profiles.rust.marker"), "Cargo.lock");
    assert_eq!(config.get_value("profiles.rust.type"), "command");
}

/// @covers: DefaultConfigProvider::get_value
#[test]
fn test_get_value_missing_key_returns_empty() {
    let tmp = tempfile::tempdir().unwrap();
    let config_path = create_test_config(tmp.path());

    let config = load_config(&config_path).unwrap();
    assert_eq!(config.get_value("profiles.nonexistent.name"), "");
}

/// @covers: DefaultConfigProvider::get_array
#[test]
fn test_get_array_parses_inline_array() {
    let tmp = tempfile::tempdir().unwrap();
    let config_path = create_test_config(tmp.path());

    let config = load_config(&config_path).unwrap();
    let arr = config.get_array("profiles.node.alt_markers");
    assert_eq!(arr, vec!["yarn.lock", "pnpm-lock.yaml"]);
}

/// @covers: DefaultConfigProvider::get_array
#[test]
fn test_get_array_missing_key_returns_empty() {
    let tmp = tempfile::tempdir().unwrap();
    let config_path = create_test_config(tmp.path());

    let config = load_config(&config_path).unwrap();
    assert!(config.get_array("profiles.rust.nonexistent").is_empty());
}

/// @covers: DefaultConfigProvider::load
#[test]
fn test_load_config_nonexistent_file_returns_error() {
    let result = load_config(Path::new("/nonexistent/path/config.toml"));
    assert!(result.is_err());
}

/// @covers: DefaultConfigProvider::load (comment with hash inside quotes)
#[test]
fn test_load_config_preserves_hash_inside_quotes() {
    let tmp = tempfile::tempdir().unwrap();
    let config_path = tmp.path().join("test.toml");
    fs::write(
        &config_path,
        "[profiles.test]\nname = \"Test\"\nmarker = \"test.lock\"\ntype = \"command\"\npattern = 'Removed #(\\d+) files'\n",
    )
    .unwrap();

    let config = load_config(&config_path).unwrap();
    assert_eq!(
        config.get_value("profiles.test.pattern"),
        r"Removed #(\d+) files"
    );
}

// ─── CleanProfile Integration ───────────────────────────────────────────────

/// @covers: load_profile
#[test]
fn test_load_profile_command_type_fields() {
    let tmp = tempfile::tempdir().unwrap();
    let config_path = create_test_config(tmp.path());
    let config = load_config(&config_path).unwrap();

    let p = load_profile("rust", &config);
    assert_eq!(p.key, "rust");
    assert_eq!(p.name, "Rust (Cargo)");
    assert_eq!(p.marker, "Cargo.lock");
    assert_eq!(p.profile_type, "command");
    assert_eq!(p.command, "cargo clean");
    assert_eq!(p.clean_dir, "target");
}

/// @covers: load_profile
#[test]
fn test_load_profile_remove_type_fields() {
    let tmp = tempfile::tempdir().unwrap();
    let config_path = create_test_config(tmp.path());
    let config = load_config(&config_path).unwrap();

    let p = load_profile("node", &config);
    assert_eq!(p.profile_type, "remove");
    assert!(p.targets.contains(&"node_modules".to_string()));
    assert!(p.optional_targets.contains(&".next".to_string()));
}

/// @covers: CleanProfile::all_markers
#[test]
fn test_all_markers_includes_primary_and_alternates() {
    let tmp = tempfile::tempdir().unwrap();
    let config_path = create_test_config(tmp.path());
    let config = load_config(&config_path).unwrap();

    let p = load_profile("node", &config);
    let markers = p.all_markers();
    assert!(markers.contains(&"package-lock.json"));
    assert!(markers.contains(&"yarn.lock"));
    assert!(markers.contains(&"pnpm-lock.yaml"));
}

// ─── ProjectScanner Integration ─────────────────────────────────────────────

/// @covers: DefaultProjectScanner::scan_for_projects
#[test]
fn test_scan_for_projects_finds_all_by_marker() {
    let tmp = tempfile::tempdir().unwrap();

    // Create 3 node projects
    for name in &["app-web", "app-api", "lib-shared"] {
        write_file(tmp.path(), &format!("{name}/package-lock.json"), "");
        write_file(tmp.path(), &format!("{name}/node_modules/dep.js"), "module");
    }

    let config_path = create_test_config(tmp.path());
    let config = load_config(&config_path).unwrap();
    let profile = load_profile("node", &config);

    let scanner = create_scanner(tmp.path().to_path_buf(), vec![], vec![], false, false);
    let found = scanner.scan_for_projects(&profile.all_markers());
    assert_eq!(found.len(), 3);
}

/// @covers: DefaultProjectScanner::scan_for_projects
#[test]
fn test_scan_for_projects_returns_empty_for_no_matches() {
    let tmp = tempfile::tempdir().unwrap();
    write_file(tmp.path(), "something.txt", "hello");

    let config_path = create_test_config(tmp.path());
    let config = load_config(&config_path).unwrap();
    let profile = load_profile("rust", &config);

    let scanner = create_scanner(tmp.path().to_path_buf(), vec![], vec![], false, false);
    let found = scanner.scan_for_projects(&profile.all_markers());
    assert_eq!(found.len(), 0);
}

/// @covers: DefaultProjectScanner::filter_projects
#[test]
fn test_filter_projects_applies_exclude_patterns() {
    let tmp = tempfile::tempdir().unwrap();

    for name in &["app-web", "app-api", "lib-shared"] {
        write_file(tmp.path(), &format!("{name}/package-lock.json"), "");
    }

    let config_path = create_test_config(tmp.path());
    let config = load_config(&config_path).unwrap();
    let profile = load_profile("node", &config);

    let scanner = create_scanner(
        tmp.path().to_path_buf(),
        vec!["lib-shared".to_string()],
        vec![],
        false,
        false,
    );
    let found = scanner.scan_for_projects(&profile.all_markers());
    let (to_process, skipped) = scanner.filter_projects(&found);

    assert_eq!(to_process.len(), 2);
    assert_eq!(skipped.len(), 1);
}

/// @covers: DefaultProjectScanner::filter_projects
#[test]
fn test_filter_projects_applies_include_patterns() {
    let tmp = tempfile::tempdir().unwrap();

    for name in &["app-web", "app-api", "lib-shared"] {
        write_file(tmp.path(), &format!("{name}/package-lock.json"), "");
    }

    let config_path = create_test_config(tmp.path());
    let config = load_config(&config_path).unwrap();
    let profile = load_profile("node", &config);

    let scanner = create_scanner(
        tmp.path().to_path_buf(),
        vec![],
        vec!["app-web".to_string()],
        false,
        false,
    );
    let found = scanner.scan_for_projects(&profile.all_markers());
    let (to_process, skipped) = scanner.filter_projects(&found);

    assert_eq!(to_process.len(), 1);
    assert_eq!(skipped.len(), 2);
}

/// @covers: DefaultProjectScanner::filter_projects
#[test]
fn test_filter_projects_clean_all_ignores_exclude() {
    let tmp = tempfile::tempdir().unwrap();

    for name in &["app-web", "app-api", "lib-shared"] {
        write_file(tmp.path(), &format!("{name}/package-lock.json"), "");
    }

    let config_path = create_test_config(tmp.path());
    let config = load_config(&config_path).unwrap();
    let profile = load_profile("node", &config);

    let scanner = create_scanner(
        tmp.path().to_path_buf(),
        vec!["lib-shared".to_string()],
        vec![],
        true, // clean_all
        false,
    );
    let found = scanner.scan_for_projects(&profile.all_markers());
    let (to_process, skipped) = scanner.filter_projects(&found);

    assert_eq!(to_process.len(), 3);
    assert_eq!(skipped.len(), 0);
}

// ─── Utility Functions ──────────────────────────────────────────────────────

/// @covers: format_size
#[test]
fn test_format_size_bytes() {
    assert_eq!(format_size(0), "0 B");
    assert_eq!(format_size(512), "512 B");
}

/// @covers: format_size
#[test]
fn test_format_size_kib() {
    assert_eq!(format_size(1024), "1.00 KiB");
    assert_eq!(format_size(1536), "1.50 KiB");
}

/// @covers: format_size
#[test]
fn test_format_size_mib() {
    assert_eq!(format_size(1024 * 1024), "1.00 MiB");
}

/// @covers: format_size
#[test]
fn test_format_size_gib() {
    assert_eq!(format_size(1024 * 1024 * 1024), "1.00 GiB");
}

/// @covers: dir_size_bytes
#[test]
fn test_dir_size_bytes_nonexistent_returns_zero() {
    assert_eq!(dir_size_bytes(Path::new("/nonexistent/path/12345")), 0);
}

/// @covers: dir_size_bytes
#[test]
fn test_dir_size_bytes_measures_directory_content() {
    let tmp = tempfile::tempdir().unwrap();
    let data = vec![0u8; 4096];
    let file_path = tmp.path().join("data.bin");
    {
        let mut f = fs::File::create(&file_path).unwrap();
        f.write_all(&data).unwrap();
        // f drops here, flushing and closing the file handle
    }

    let size = dir_size_bytes(tmp.path());
    assert_eq!(size, 4096);
}

/// @covers: DefaultProjectScanner::relative_path
#[test]
fn test_relative_path_strips_search_root() {
    let tmp = tempfile::tempdir().unwrap();
    let scanner = create_scanner(tmp.path().to_path_buf(), vec![], vec![], false, false);

    let full = tmp.path().join("app-web");
    let rel = scanner.relative_path(&full);
    assert_eq!(rel.to_string_lossy(), "app-web");
}

File: java/tests/config/ConfigIntTest.java
package tests.config;

import main.features.config.saf.ConfigFactory;

import java.io.File;
import java.io.FileOutputStream;
import java.io.FileWriter;
import java.io.IOException;

/**
 * Integration tests for config feature.
 * Tests TOML parsing, profile loading, project scanning, filtering, size utilities.
 */
public class ConfigIntTest {

    static int passed = 0;
    static int failed = 0;

    public static void main(String[] args) {
        System.out.println("=== Config Integration Tests ===");
        System.out.println();

        testLoadConfigReadsAllProfiles();
        testGetValueReturnsUnquotedString();
        testGetValueMissingReturnsEmpty();
        testGetArrayParsesInlineArray();
        testGetArrayMissingReturnsEmpty();
        testLoadConfigPreservesHashInsideQuotes();
        testLoadProfileCommandTypeFields();
        testLoadProfileRemoveTypeFields();
        testAllMarkersIncludesPrimaryAndAlternates();
        testScanForProjectsFindsByMarker();
        testScanForProjectsReturnsEmptyForNoMatch();
        testFilterProjectsAppliesExclude();
        testFilterProjectsAppliesInclude();
        testFilterProjectsCleanAllIgnoresExclude();
        testFormatSizeBytes();
        testFormatSizeKib();
        testFormatSizeMib();
        testFormatSizeGib();
        testDirSizeBytesNonexistentReturnsZero();
        testDirSizeBytesMeasuresContent();
        testRelativePathStripsSearchRoot();

        System.out.println();
        System.out.println("Results: " + passed + " passed, " + failed + " failed");
        if (failed > 0) System.exit(1);
    }

    // ─── Helpers ────────────────────────────────────────────────────────────

    static String createTempDir() {
        var dir = System.getProperty("java.io.tmpdir") + File.separator
                + "dc-test-" + System.nanoTime();
        new File(dir).mkdirs();
        return dir;
    }

    static void writeFile(String root, String relative, String content) {
        var f = new File(root + File.separator + relative);
        f.getParentFile().mkdirs();
        try { var w = new FileWriter(f); w.write(content); w.close(); }
        catch (IOException e) { throw new RuntimeException(e); }
    }

    static void writeBinary(String root, String relative, int size) {
        var f = new File(root + File.separator + relative);
        f.getParentFile().mkdirs();
        try { var out = new FileOutputStream(f); out.write(new byte[size]); out.close(); }
        catch (IOException e) { throw new RuntimeException(e); }
    }

    static void deleteRecursive(File f) {
        if (f.isDirectory()) {
            var children = f.listFiles();
            if (children != null) for (var c : children) deleteRecursive(c);
        }
        f.delete();
    }

    static String writeTestConfig(String dir) {
        var path = dir + File.separator + "profiles.toml";
        writeFile(dir, "profiles.toml",
            "[settings]\ndefault_profiles = [\"rust\"]\n\n"
            + "[profiles.rust]\nname = \"Rust (Cargo)\"\nmarker = \"Cargo.lock\"\n"
            + "type = \"command\"\ncommand = \"cargo clean\"\nclean_dir = \"target\"\n"
            + "source_extensions = [\".rs\", \".toml\"]\nsearch_exclude = [\"target\"]\n\n"
            + "[profiles.node]\nname = \"Node.js\"\nmarker = \"package-lock.json\"\n"
            + "alt_markers = [\"yarn.lock\", \"pnpm-lock.yaml\"]\ntype = \"remove\"\n"
            + "targets = [\"node_modules\"]\noptional_targets = [\".next\", \"dist\"]\n"
            + "source_extensions = [\".js\", \".ts\"]\nsearch_exclude = [\"node_modules\"]\n");
        return path;
    }

    static void check(String name, boolean condition) {
        if (condition) { passed++; System.out.println("  PASS  " + name); }
        else { failed++; System.out.println("  FAIL  " + name); }
    }

    // ─── Tests ──────────────────────────────────────────────────────────────

    static void testLoadConfigReadsAllProfiles() {
        var dir = createTempDir();
        try {
            var config = ConfigFactory.loadConfig(writeTestConfig(dir));
            check("test_load_config_reads_all_profiles", config.profileKeys().length == 2);
        } finally { deleteRecursive(new File(dir)); }
    }

    static void testGetValueReturnsUnquotedString() {
        var dir = createTempDir();
        try {
            var config = ConfigFactory.loadConfig(writeTestConfig(dir));
            check("test_get_value_returns_unquoted_string",
                    config.getValue("profiles.rust.name").equals("Rust (Cargo)")
                    && config.getValue("profiles.rust.marker").equals("Cargo.lock"));
        } finally { deleteRecursive(new File(dir)); }
    }

    static void testGetValueMissingReturnsEmpty() {
        var dir = createTempDir();
        try {
            var config = ConfigFactory.loadConfig(writeTestConfig(dir));
            check("test_get_value_missing_returns_empty",
                    config.getValue("profiles.nonexistent.name").isEmpty());
        } finally { deleteRecursive(new File(dir)); }
    }

    static void testGetArrayParsesInlineArray() {
        var dir = createTempDir();
        try {
            var config = ConfigFactory.loadConfig(writeTestConfig(dir));
            var arr = config.getArray("profiles.node.alt_markers");
            check("test_get_array_parses_inline_array",
                    arr.length == 2 && arr[0].equals("yarn.lock") && arr[1].equals("pnpm-lock.yaml"));
        } finally { deleteRecursive(new File(dir)); }
    }

    static void testGetArrayMissingReturnsEmpty() {
        var dir = createTempDir();
        try {
            var config = ConfigFactory.loadConfig(writeTestConfig(dir));
            check("test_get_array_missing_returns_empty",
                    config.getArray("profiles.rust.nonexistent").length == 0);
        } finally { deleteRecursive(new File(dir)); }
    }

    static void testLoadConfigPreservesHashInsideQuotes() {
        var dir = createTempDir();
        try {
            writeFile(dir, "test.toml",
                    "[profiles.test]\nname = \"Test\"\nmarker = \"test.lock\"\n"
                    + "type = \"command\"\npattern = 'Removed #(\\d+) files'\n");
            var config = ConfigFactory.loadConfig(dir + File.separator + "test.toml");
            check("test_load_config_preserves_hash_inside_quotes",
                    config.getValue("profiles.test.pattern").equals("Removed #(\\d+) files"));
        } finally { deleteRecursive(new File(dir)); }
    }

    static void testLoadProfileCommandTypeFields() {
        var dir = createTempDir();
        try {
            var config = ConfigFactory.loadConfig(writeTestConfig(dir));
            var p = ConfigFactory.loadProfile("rust", config);
            check("test_load_profile_command_type_fields",
                    p.key.equals("rust") && p.name.equals("Rust (Cargo)")
                    && p.profileType.equals("command") && p.command.equals("cargo clean"));
        } finally { deleteRecursive(new File(dir)); }
    }

    static void testLoadProfileRemoveTypeFields() {
        var dir = createTempDir();
        try {
            var config = ConfigFactory.loadConfig(writeTestConfig(dir));
            var p = ConfigFactory.loadProfile("node", config);
            check("test_load_profile_remove_type_fields",
                    p.profileType.equals("remove") && p.targets[0].equals("node_modules"));
        } finally { deleteRecursive(new File(dir)); }
    }

    static void testAllMarkersIncludesPrimaryAndAlternates() {
        var dir = createTempDir();
        try {
            var config = ConfigFactory.loadConfig(writeTestConfig(dir));
            var p = ConfigFactory.loadProfile("node", config);
            var markers = p.allMarkers();
            check("test_all_markers_includes_primary_and_alternates",
                    markers.length == 3 && markers[0].equals("package-lock.json")
                    && markers[1].equals("yarn.lock"));
        } finally { deleteRecursive(new File(dir)); }
    }

    static void testScanForProjectsFindsByMarker() {
        var dir = createTempDir();
        try {
            writeFile(dir, "app-web/package-lock.json", "");
            writeFile(dir, "app-api/package-lock.json", "");
            writeFile(dir, "lib/package-lock.json", "");
            var config = ConfigFactory.loadConfig(writeTestConfig(dir));
            var profile = ConfigFactory.loadProfile("node", config);
            var scanner = ConfigFactory.createScanner(dir, new String[0], new String[0], false, false);
            var found = scanner.scanForProjects(profile.allMarkers());
            check("test_scan_for_projects_finds_by_marker", found.length == 3);
        } finally { deleteRecursive(new File(dir)); }
    }

    static void testScanForProjectsReturnsEmptyForNoMatch() {
        var dir = createTempDir();
        try {
            writeFile(dir, "something.txt", "hello");
            var config = ConfigFactory.loadConfig(writeTestConfig(dir));
            var profile = ConfigFactory.loadProfile("rust", config);
            var scanner = ConfigFactory.createScanner(dir, new String[0], new String[0], false, false);
            var found = scanner.scanForProjects(profile.allMarkers());
            check("test_scan_for_projects_returns_empty_for_no_match", found.length == 0);
        } finally { deleteRecursive(new File(dir)); }
    }

    static void testFilterProjectsAppliesExclude() {
        var dir = createTempDir();
        try {
            writeFile(dir, "app-web/package-lock.json", "");
            writeFile(dir, "app-api/package-lock.json", "");
            writeFile(dir, "lib-shared/package-lock.json", "");
            var config = ConfigFactory.loadConfig(writeTestConfig(dir));
            var profile = ConfigFactory.loadProfile("node", config);
            var scanner = ConfigFactory.createScanner(dir, new String[]{"lib-shared"}, new String[0], false, false);
            var found = scanner.scanForProjects(profile.allMarkers());
            var filtered = scanner.filterProjects(found);
            check("test_filter_projects_applies_exclude",
                    filtered[0].length == 2 && filtered[1].length == 1);
        } finally { deleteRecursive(new File(dir)); }
    }

    static void testFilterProjectsAppliesInclude() {
        var dir = createTempDir();
        try {
            writeFile(dir, "app-web/package-lock.json", "");
            writeFile(dir, "app-api/package-lock.json", "");
            writeFile(dir, "lib-shared/package-lock.json", "");
            var config = ConfigFactory.loadConfig(writeTestConfig(dir));
            var profile = ConfigFactory.loadProfile("node", config);
            var scanner = ConfigFactory.createScanner(dir, new String[0], new String[]{"app-web"}, false, false);
            var found = scanner.scanForProjects(profile.allMarkers());
            var filtered = scanner.filterProjects(found);
            check("test_filter_projects_applies_include",
                    filtered[0].length == 1 && filtered[1].length == 2);
        } finally { deleteRecursive(new File(dir)); }
    }

    static void testFilterProjectsCleanAllIgnoresExclude() {
        var dir = createTempDir();
        try {
            writeFile(dir, "app-web/package-lock.json", "");
            writeFile(dir, "lib-shared/package-lock.json", "");
            var config = ConfigFactory.loadConfig(writeTestConfig(dir));
            var profile = ConfigFactory.loadProfile("node", config);
            var scanner = ConfigFactory.createScanner(dir, new String[]{"lib-shared"}, new String[0], true, false);
            var found = scanner.scanForProjects(profile.allMarkers());
            var filtered = scanner.filterProjects(found);
            check("test_filter_projects_clean_all_ignores_exclude",
                    filtered[0].length == 2 && filtered[1].length == 0);
        } finally { deleteRecursive(new File(dir)); }
    }

    static void testFormatSizeBytes() {
        check("test_format_size_bytes",
                ConfigFactory.formatSize(0).equals("0 B") && ConfigFactory.formatSize(512).equals("512 B"));
    }

    static void testFormatSizeKib() {
        check("test_format_size_kib", ConfigFactory.formatSize(1024).equals("1.00 KiB"));
    }

    static void testFormatSizeMib() {
        check("test_format_size_mib", ConfigFactory.formatSize(1024 * 1024).equals("1.00 MiB"));
    }

    static void testFormatSizeGib() {
        check("test_format_size_gib", ConfigFactory.formatSize(1024L * 1024L * 1024L).equals("1.00 GiB"));
    }

    static void testDirSizeBytesNonexistentReturnsZero() {
        check("test_dir_size_bytes_nonexistent_returns_zero",
                ConfigFactory.dirSizeBytes("/nonexistent/path/12345") == 0);
    }

    static void testDirSizeBytesMeasuresContent() {
        var dir = createTempDir();
        try {
            writeBinary(dir, "data.bin", 4096);
            check("test_dir_size_bytes_measures_content", ConfigFactory.dirSizeBytes(dir) == 4096);
        } finally { deleteRecursive(new File(dir)); }
    }

    static void testRelativePathStripsSearchRoot() {
        var dir = createTempDir();
        try {
            var scanner = ConfigFactory.createScanner(dir, new String[0], new String[0], false, false);
            var rel = scanner.relativePath(dir + File.separator + "app-web");
            check("test_relative_path_strips_search_root", rel.equals("app-web"));
        } finally { deleteRecursive(new File(dir)); }
    }
}
